### PR TITLE
[AIRFLOW-2973] Add Python 3.6 to Supported Prog Langs

### PR DESCRIPTION
**Pull request by kaxil**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-2973
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
I remember Fokko and Joy had tested Airflow on Python 3.6. Also, DigitalOcean has also been using Airflow with 3.6 as mentioned in the mailing list: https://lists.apache.org/thread.html/4b3ab976ad268f3da96c5317290d11120c26c7e0f92886eef5b8b50f@%3Cdev.airflow.apache.org%3E

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by kaxil** [comment]
cc @bolkedebruin @Fokko

**Comment by ashb** [comment]
Is it worth adding/changing Py 3.5 to Py 3.6 in Travis? (I don't think we want to have both py 3.5 and py 3.6 tests as not that much has changed and we don't want _more_ runtime) but...?

**Review by ashb — APPROVED** [review]

**Comment by bolkedebruin** [comment]
Agree with @ashb

**Comment by kaxil** [comment]
@bolkedebruin @ashb Do you guys suggest to add Python 3.4 and Python 3.6 both to Travis? As currently just have py27 and py35 on Travis.

**Comment by Fokko** [comment]
This would give too much load on Travis. Apache infra paying for cpu hours.

**Comment by kaxil** [comment]
@Fokko Agree with that, so do you like to keep it as it is or changing py3.5 to py3.6 and keep py2.7 and py3.6 ?

**Comment by bolkedebruin** [comment]
2.7 and 3.6 (and maybe 3.7 instead)

**Comment by tedmiston** [comment]
@kaxil @Fokko I also put together a quick complementary PR https://github.com/apache/incubator-airflow/pull/3816 adding / upgrading to 3.6 in all of the places.  CI is still running so I may expect some test failures but hopefully nothing major.  I would be happy to merge this into Kaxil's current PR or as a follow-on PR.

Personally I would prefer Travis to run against 3.6 over 3.5.  It's been out for a while now, and AFAIK we don't have a compelling reason to prefer 3.5 today.




That said it would be nice if we could have some sort of nightly-ish build to run on all supported subversions for a very thorough test.

**Review by tedmiston — APPROVED** [review]

**Comment by kaxil** [comment]
@tedmiston I have merged the changes to this PR but as a follow-up to this PR, I suppose it would be great if you can test Python 3.7 and as Bolke mentioned, we can then replace 3.6 with 3.7 :)

**Comment by kaxil** [comment]
The way CI works now with Docker, we will have to change bits on our Docker CI Image as well. I have pinged @gerardo on the Dockerised PR to help the case where we would have to run tests for multiple python versions. I am hence just keeping the scope of this PR to add 3.6 to setup.py, or we can close this PR entirely and have first sort out our CI pipeline and can add these changes over there? 

cc @ashb @Fokko

**Comment by gerardo** [comment]
(posting again in here)

@kaxil give we're using tox for this, it should be a matter of just adding/updating the runtimes inside the tox file [here](https://github.com/apache/incubator-airflow/blob/master/tox.ini#L20) and [here](https://github.com/apache/incubator-airflow/blob/master/tox.ini#L40) and then pass the desired python version through the Travis script as we do now.

**Comment by Fokko** [comment]
@gerardo This does not make sense to me. Why are we installing both 2 and 3 in the ci image? https://github.com/apache/incubator-airflow-ci/blob/master/Dockerfile

Ideally we would have a Python 2 and Python 3 image, but this can become complicated quickly.

**Comment by gerardo** [comment]
@Fokko The only reason 2 versions are being installed is that I wanted to follow the CI process as close as it was [before merging the docker-ci changes](https://travis-ci.org/apache/incubator-airflow/builds/416219284). At the time, the tox target and the python version to be used were being matched (run a python3 tox target and run tox under python3), although I never understood the reason. We only use the system python to install tox.

I only remember coming across a single issue with a `PythonVirtualOperator` test failing [because of a version mismatch](
https://github.com/apache/incubator-airflow/pull/3393#discussion_r189708965) (tox running python3 and `PythonVirtualOperator` trying to use the system python, which could be python2), but if that happens, it should be simple to fix.

**Comment by kaxil** [comment]
@gerado I already tried changing that, it says python3.6 interpreter not found.

Here is the link to the Travis job:

https://travis-ci.org/apache/incubator-airflow/builds/421714589

**Comment by gerardo** [comment]
@kaxil mmm, it seems it needs to be installed in the system to be used: https://tox.readthedocs.io/en/latest/example/general.html#basepython-defaults-overriding

And it seems the only python3 version available in Ubuntu Xenial is 3.5.1: https://packages.ubuntu.com/search?keywords=python3-dev&searchon=names&suite=xenial&section=all

It might be a matter of having a separate image (like @Fokko suggested) using a more recent Ubuntu (`Artful` or `Bionic`). The official [python docker images](https://hub.docker.com/_/python/) are also based in Debian, that's another option.

**Comment by kaxil** [comment]
Would you be able to do that if you have got time?

**Comment by gerardo** [comment]
@kaxil sure! I have time this week to work on this.

**Comment by Fokko** [comment]
@gerardo On the `incubator-airflow-ci` we can create two branches, one for Python 2 and one for Python 3. I can configure the docker-hub to publish two images `incubator-airflow-ci:2` and `incubator-airflow-ci:3` for example

**Comment by tedmiston** [comment]
@kaxil Sure.  Btw, I've started a 3.7 test branch currently living here - https://github.com/astronomerio/incubator-airflow/tree/python-3.7-experimental.

It looks like the merge of #3816 into here yesterday was reverted.  It sounds like this PR may be on hold until the Docker image is reworked to support multiple subversions of Python 3.  Is that correct?

@gerardo Can you tag me in that PR ^ when you're done?  I'd like to make a similar change in my follow up PR to support Python 3.7 (perhaps with an optional flag for now so we don't burn Apache Infra CI time).

**Comment by kaxil** [comment]
That's awesome @gerardo . Looking forward for it.

@tedmiston Great. Would love to see Py3.7 PR, I think #3816 can be converted to be a 3.7 Upgrade PR. So once 3.6 is fixed, you can probably overwrite it with 3.7.

**Comment by gerardo** [comment]
> @kaxil sure! I have time this week to work on this.

BTW, I hugely underestimated the amount of free time I had available 😂

Here's the PR that creates separate images for python2 and python3: https://github.com/apache/incubator-airflow-ci/pull/4

@tedmiston Regarding python3.7, it's still not available to any Ubuntu version, not even through a PPA. Even [cosmic cuttlefish](https://wiki.ubuntu.com/CosmicCuttlefish), which hasn't been released yet, supports only [3.6](https://packages.ubuntu.com/cosmic/python3).

**Comment by Fokko** [comment]
@gerardo When updating the CI, we run into problems since we need a to point the Python3 tests to the Python 3 image: https://travis-ci.org/Fokko/incubator-airflow

I think we need to get rid of the `python3` and `python3.6` command and generalize everything to just `python`, and pick the correct base images if we run on either `python2` or `python3`.

I think updating to Python 3.7 doesn't make sense since a lot of companies are still on 3.6 (or even lower :-)

**Comment by gerardo** [comment]
@Fokko @kaxil @tedmiston 
I've done some more work on https://github.com/apache/incubator-airflow-ci/pull/4, but I've found a few issues with the Python3.6 upgrade, plus other issues with some assumptions that the test suite makes:

- [`tests.operators.test_virtualenv_operator.test_python_3`](https://github.com/apache/incubator-airflow/blob/6a7f3887460c91bc4da90fe8d31836a225324a1a/tests/operators/test_virtualenv_operator.py#L124) fails when run with the `python2` image ([logs](https://travis-ci.com/gerardo/incubator-airflow/jobs/157826745#L4443)). Using `skipIf` to check what python version is present should be enough. And maybe we should do the same with [test_python_2 and test_python_2.7](https://github.com/apache/incubator-airflow/blob/6a7f3887460c91bc4da90fe8d31836a225324a1a/tests/operators/test_virtualenv_operator.py#L124)
- [tests.operators.test_virtualenv_operator.test_string_args](https://github.com/apache/incubator-airflow/blob/6a7f3887460c91bc4da90fe8d31836a225324a1a/tests/operators/test_virtualenv_operator.py#L170) is a weird test. It uses a method called `_invert_python_major_version()` to use the opposite version to the one available, which is causing problems with [python2 tests](https://travis-ci.com/gerardo/incubator-airflow/jobs/157826745#L4503). Do you guys understand this test?
- There was a change between `python3.5` and `python3.6` in the error message emitted when an object is not serializable ([logs](https://travis-ci.com/gerardo/incubator-airflow/jobs/157826750#L4568)). It's just a matter of changing the test to check for the new message format.
- One of the tz tests is failing as well, more specifically, `test_following_previous_schedule_daily_dag_CEST_to_CET` ([logs](https://travis-ci.com/gerardo/incubator-airflow/jobs/157826750#L4367)). I'm not sure where to start. It seems to happen only with `python3.6`.

**Comment by Fokko** [comment]
Give me some time to look into this.

File: setup.py
Comment (bryantbiggs) [inline setup.py:399]: It looks like this line should be dropped since 3.4 is not tested in the CI process

**Comment by holdenk** [comment]
Hi @gerardo I did some looking into tests.operators.test_virtualenv_operator.test_string_args  as well. Looking at the Python operator description string args is the only supported method of allowing configuration when the Python versions don't match between airflow and operator -- which would be why it does the test through these string arguments. 

It looks like if we want to make that test workable in CI then the Python 2 image needs Python 3 installed as well and vice versa just different default Pythons configured.

File: setup.py
Comment (bryantbiggs) [reply setup.py:399]: disregard - this has been fixed in https://github.com/apache/incubator-airflow/pull/4259

**Comment by Fokko** [comment]
Already the case https://github.com/apache/airflow/blob/master/setup.py#L429

👍

**Pull request closed without merging**
